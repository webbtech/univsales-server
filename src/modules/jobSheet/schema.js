import { gql } from 'apollo-server-lambda'

export default gql`
extend type Mutation {
  jobSheetDuplicate(id: ID!): JobSheet
  jobSheetPersist(addressInput: AddressInput, jobSheetInput: JobSheetInput!): JobSheet
  jobSheetPersistFeatures(features: String!, id: ID!): JobSheet
  jobSheetPersistGroup(input: GroupInput!): JobSheetGroup
  jobSheetPersistOther(input: OtherInput!): JobSheetOther
  jobSheetPersistWindow(input: WindowInput!): JobSheetWindow
  jobSheetRemove(id: ID!): DBResult
  jobSheetRemoveGroup(id: ID!): DBResult
  jobSheetRemoveOther(id: ID!): DBResult
  jobSheetRemoveWindow(id: ID!): DBResult
}
extend type Query {
  jobSheetData(jobSheetID: ID!): JobSheetData
  jobSheetGroup(groupID: ID!): JobSheetGroup
  jobSheetOther(otherID: ID!): JobSheetOther
  jobSheetWindow(windowID: ID!): JobSheetWindow
  searchJobSheetsByCustomer(customerID: ID!): [JobSheet]
}

#
## ================================== Types ==================================
#

type GroupItemCosts {
  extendTotal: Float
  extendUnit: Float
}

type GroupItemDims {
  height: WindowDims
  width: WindowDims
}

type GroupItemSpecs {
  extendSqft: Int
  options: String
  overSize: Int
  sqft: Int
}

type GroupWindowDims {
  decimal: Float
  diff: Float
  fraction: String
  inch: Int
}

type JobSheet {
  _id: ID!
  addressID: Address
  duplicateNumber: Int
  createdAt: String
  customerID: Customer
  features: String
  number: Int!
  updatedAt: String
}

type JobSheetData {
  groups: [JobSheetGroup]
  jobsheet: JobSheet
  other: [JobSheetOther]
  windows: [JobSheetWindow]
}

type JobSheetGroup {
  _id: ID!
  costs: JobSheetGroupCosts
  createdAt: String
  dims: JobSheetGroupDims
  items: [JobSheetGroupItem]
  jobsheetID: ID!
  qty: Int
  rooms: [String]
  specs: JobSheetGroupSpecs
  updatedAt: String
}

type JobSheetGroupCosts {
  discounted: Float
  discountedAmount: Float
  extendTotal: Float
  extendUnit: Float
  install: Float
  installType: Float
  netUnit: Float
  options: Float
  trim: Float
  windows: Float
}

type JobSheetGroupDims {
  height: GroupWindowDims
  width: GroupWindowDims
}

type JobSheetGroupItem {
  _id: ID
  costs: GroupItemCosts
  dims: GroupItemDims
  product: Product
  productID: ID
  qty: Int
  specs: GroupItemSpecs
}

type JobSheetGroupSpecs {
  groupTypeDescription: String
  installType: String
  options: String
  optionsDoc: OptionsDoc
  sqft: Int
  style: String
  trim: String
}

type JobSheetItemCosts {
  discounted: Float
  discountedAmount: Float
  extendTotal: Float
  extendUnit: Float
  install: Float
  installType: Float
  netUnit: Float
  options: Float
  trim: Float
  window: Float
}

type JobSheetOther {
  _id: ID!
  costs: OtherCosts
  createdAt: String
  description: String
  jobsheetID: ID
  product: String
  qty: Int
  rooms: [String]
  specs: OtherSpecs
  updatedAt: String
}

type JobSheetWindow {
  _id: ID!
  brickmould: BrickMould
  costs: JobSheetItemCosts
  createdAt: String
  dims: JobSheetWindowDims
  jobsheetID: ID
  productID: Product
  qty: Int
  rooms: [String]
  specs: JobSheetWindowSpecs
  updatedAt: String
}

type JobSheetWindowDims {
  height: WindowDims
  width: WindowDims
}

type JobSheetWindowSpecs {
  installType: String
  options: String
  optionsDoc: OptionsDoc
  overSize: Int
  sqft: Int
  trim: String
}

type OtherCosts {
  extendTotal: Float
  extendUnit: Float
}

type OtherSpecs {
  location: String
  options: String
}

type WindowDims {
  decimal: Float
  fraction: String
  inch: Int
  overSize: Int
  round: Int
  underSize: Int
}

type OptionsDoc {
  color: String
  customColor: String
  darker: Boolean
  extraDetails: String
  window: [String]
}

type BrickMould {
  dims: BrickMouldDims
  size: Int
  specs: String
}

type BrickMouldDim {
  decimal: Float
  inch: Int
  fraction: String
}

type BrickMouldDims {
  height: BrickMouldDim
  width: BrickMouldDim
}

#
## ================================== Inputs ==================================
#

input GroupCostsInput {
  discountAmount: Float
  discounted: Float
  extendTotal: Float
  extendUnit: Float
  install: Float
  installType: Float
  netUnit: Float
  options: Float
  trim: Float
  windows: Float
}

input GroupDimsInput {
  decimal: Float
  diff: Float
  fraction: String
  inch: Int
}

input GroupHWDimsInput {
  height: GroupDimsInput
  width: GroupDimsInput
}

input GroupInput {
  _id: ID
  costs: GroupCostsInput
  dims: GroupHWDimsInput
  items: [GroupItemInput]
  jobsheetID: ID!
  qty: Int
  rooms: [String]
  specs: GroupSpecsInput
}

input GroupItemCostsInput {
  extendTotal: Float
  extendUnit: Float
}

input GroupItemInput {
  _id: ID!
  costs: GroupItemCostsInput
  dims: WindowHWDimsInput
  product: GroupItemProductInput
  productID: ID!
  qty: Int
  specs: GroupItemSpecsInput
}

input GroupItemProductInput {
  name: String
}

input GroupItemSpecsInput {
  extendSqft: Int
  options: String
  overSize: Int
  sqft: Int
}

input GroupSpecsInput {
  groupTypeDescription: String
  installType: String
  options: String
  optionsDoc: OptionsDocInput
  sqft: Int
  style: String
  trim: String
}

input JobSheetInput {
  _id: ID
  customerID: ID!
  features: String
}

input OtherCostsInput {
  extendTotal: Float
  extendUnit: Float
}

input OtherInput {
  _id: ID
  costs: OtherCostsInput
  description: String
  jobsheetID: ID
  product: String
  qty: Int
  rooms: [String]
  specs: OtherSpecsInput
}

input OtherSpecsInput {
  location: String
  options: String
}

input WindowCostsInput {
  discounted: Float
  discountedAmount: Float
  extendTotal: Float
  extendUnit: Float
  install: Float
  installType: Float
  netUnit: Float
  options: Float
  trim: Float
  window: Float
}

input WindowDimsInput {
  decimal: Float
  fraction: String
  inch: Int
  overSize: Int
  round: Int
  underSize: Int
}

input WindowHWDimsInput {
  height: WindowDimsInput
  width: WindowDimsInput
}

input WindowInput {
  _id: ID
  costs: WindowCostsInput
  dims: WindowHWDimsInput
  jobsheetID: ID!
  productID: ID!
  qty: Int
  rooms: [String]
  specs: WindowSpecsInput
}

input WindowSpecsInput {
  extendSqft: Int
  installType: String
  options: String
  optionsDoc: OptionsDocInput
  overSize: Int
  sqft: Int
  trim: String
}

input OptionsDocInput {
  color: String
  customColor: String
  darker: Boolean
  extraDetails: String
  window: [String]
}

input BrickMouldInput {
  dims: BrickMouldDimsInput
  size: Int
  specs: String
}
input BrickMouldDimsInput {
  height: BrickMouldDimInput
  width: BrickMouldDimInput
}

input BrickMouldDimInput {
  decimal: Float
  fraction: String
  inch: Int
}
`
