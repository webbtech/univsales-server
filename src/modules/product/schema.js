import { gql } from 'apollo-server'

export default gql`
extend type Mutation {
  productPersist(input: ProductInput!): Product
}

extend type Query {
  product(productID: ID!): Product
  products: [Product]
}

type Product {
  _id: ID!
  maxHeight: Int
  maxWidth: Int
  minHeight: Int
  minWidth: Int
  name: String!
  premium: ProductPremium
  sizeCost: JSON
}

type ProductPremium {
  cost: Int
  oversizeLimit: Int
}

input ProductInput {
_id: ID!
  maxHeight: Int!
  maxWidth: Int!
  minHeight: Int!
  minWidth: Int!
  name: String!
  premium: ProductPremiumInput!
  sizeCost: JSON!
}

input ProductPremiumInput {
  cost: Int!
  oversizeLimit: Int!
}
`
