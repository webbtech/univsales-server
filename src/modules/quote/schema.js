import { gql } from 'apollo-server'

export default gql`
extend type Mutation {
  createInvoice(id: ID!): Quote
  createPDF(input: PDFInput!): DBResult
  quotePersist(input: QuoteInput!): Quote
  quotePersistDiscount(input: QuoteDiscountInput!): Quote
  quoteRemove(id: ID!): DBResult
}

extend type Query {
  quote(quoteID: ID!): Quote
  quoteNearbyJobs(input: AddressGeoInput): [AddressGeo]
  pdfSignedURL(input: SignedURLInput!): PDFSignedURL
  searchCustomerRecent: [QuoteRecent]
  searchQuotes(closed: Boolean, invoiced: Boolean, year: String): Quotes
  searchQuotesByCustomer(customerID: ID!): Quotes
}

type PDFSignedURL {
  code: Int
  data: SignedURLData
  message: String
  status: String
  timestamp: Int
}

type Quote {
  _id: ID!
  closed: Boolean!
  createdAt: String
  customerID: Customer
  # deposit: Boolean!
  discount: QuoteDiscount
  invoiced: Boolean!
  itemCosts: QuoteItemCosts
  itemSummary: QuoteItemSummary
  items: QuoteItems
  jobsheetID: JobSheet
  number: Int!
  quotePrice: QuotePrice
  updatedAt: String
  version: Int!
}

type QuoteDiscount {
  description: String
  discount: Float
  subtotal: Float
  tax: Float
  total: Float
}

type QuoteItemCosts {
  group: Float
  other: Float
  subtotal: Float
  window: Float
}

type QuoteItemSummary {
  group: QuoteItemSummaryDetails
  other: QuoteItemSummaryDetails
  window: QuoteItemSummaryDetails
}

type QuoteItemSummaryCosts {
  extendTotal: Float
  extendUnit: Float
  netUnit: Float
}

type QuoteItemSummaryDetails {
  items: [QuoteItemSummaryItems]
  totals: QuoteItemSummaryExtendTotal
}

type QuoteItemSummaryExtendTotal {
  extendTotal: Float
}

type QuoteItemSummaryItems {
  costs: QuoteItemSummaryCosts
  description: String
  qty: Int
  rooms: String
  specs: JSON
}

type QuoteItems {
  group: [String]
  other: [String]
  window: [String]
}

type QuotePrice {
  outstanding: Float
  payments: Float
  subtotal: Float
  tax: Float
  total: Float
}

type QuoteRecent {
  _id: ID!
  address: QuoteRecentAddress
  customer: Customer
  customerID: ID
  invoiced: Boolean
  number: Int
  quotePrice: QuotePrice
  updatedAt: String
  version: Int
}

type QuoteRecentAddress {
  _id: ID
  city: String
  street1: String
}

type Quotes {
  quotes: [Quote]!
  totalInvoiced: Float
  totalOutstanding: Float
}

type SignedURLData {
  url: String
}

#
## ================================== Inputs ==================================
#

input DiscountInput {
  description: String
  discount: Float
  subtotal: Float
  tax: Float
  total: Float
}

input ItemCostsInput {
  group: Float
  other: Float
  subtotal: Float
  window: Float
}

input ItemSummaryCostsInput {
  extendTotal: Float
  extendUnit: Float
  netUnit: Float
}

input ItemSummaryDetailsInput {
  items: [ItemSummaryItemsInput]
  totals: ItemSummaryExtendTotalInput
}

input ItemSummaryExtendTotalInput {
  extendTotal: Float
}

input ItemSummaryInput {
  group: ItemSummaryDetailsInput
  other: ItemSummaryDetailsInput
  window: ItemSummaryDetailsInput
}

input ItemSummaryItemsInput {
  costs: ItemSummaryCostsInput
  description: String
  qty: Int
  rooms: String
  specs: JSON
}

input ItemsInput {
  group: [String]
  other: [String]
  window: [String]
}

input PDFInput {
  quoteID: ID!
  docType: String
}

input QuoteDiscountInput {
  _id: ID!
  discount: DiscountInput!
  quotePrice: QuotePriceInput!
}

input QuoteInput {
  _id: ID
  customerID: ID!
  discount: DiscountInput
  features: String
  itemCosts: ItemCostsInput
  itemSummary: ItemSummaryInput
  items: ItemsInput
  jobsheetID: ID!
  quotePrice: QuotePriceInput
  version: Int
}

input QuotePriceInput {
  outstanding: Float
  payments: Float
  subtotal: Float
  tax: Float
  total: Float
}

input SignedURLInput {
  number: Int!
  type: String!
  version: Int
}
`
