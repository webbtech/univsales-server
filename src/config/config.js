import fs from 'fs'
import path from 'path'
import yaml from 'js-yaml'
import AWS from 'aws-sdk'

// see https://hackernoon.com/you-should-use-ssm-parameter-store-over-lambda-env-variables-5197fc6ea45b
// for example using expiry with events

const config = {
  defaults: {},
  defaultsFN: 'defaults.yml',
  defaultsPath: '',
  ssmPath: '',
  config: {},
}

config.load = async function load() {
  try {
    this.setDefaults()
    await this.loadSSM()
    return this.config
  } catch (e) {
    throw new Error(e)
  }
}

config.setDefaultsPath = function setDefaultsPath(filename = this.defaultsFN) {
  this.defaultsPath = path.join(__dirname, filename)
  try {
    fs.accessSync(this.defaultsPath, fs.F_OK)
  } catch (e) {
    // console.error(e) // eslint-disable-line
    throw new Error(e)
  }
}

config.setDefaults = function setDefaults() {
  if (!this.defaultsPath) {
    this.setDefaultsPath()
  }
  try {
    this.defaults = yaml.safeLoad(fs.readFileSync(this.defaultsPath, 'utf8'))
  } catch (e) {
    // console.error(e) // eslint-disable-line
    throw new Error(e)
  }

  // Ensure we have ssm path
  if (this.defaults.usessm && !this.defaults.ssmPath) {
    throw new Error('Missing ssmPath in defaults')
  }

  // set stage
  this.defaults.nodeEnv = process.env.Stage ? process.env.Stage : process.env.NODE_ENV
}

config.getDefaults = function getDefaults() {
  return this.defaults
}

config.setSSMPath = function setSSMPath() {
  if (!this.defaults.ssmPath || !this.defaults.awsRegion) {
    throw new Error('Missing awsRegion, stage or ssmPath in defaults')
  }

  switch (this.defaults.nodeEnv) {
    case 'test':
    case 'dev':
    case 'development':
      this.ssmPath = `/test/${this.defaults.ssmPath}`
      break
    default:
      this.ssmPath = `/${this.defaults.nodeEnv}/${this.defaults.ssmPath}`
  }
}

config.loadSSM = async function loadSSM() {
  if (!this.defaults.usessm || !this.defaults.ssmPath) {
    console.log('loadSSM called without usessm or ssmPath variables set') // eslint-disable-line
    return false
  }
  this.setSSMPath()

  const ssm = new AWS.SSM({
    apiVersion: '2014-11-06',
    region: this.defaults.awsRegion,
  })

  const ssmParams = {
    Path: this.ssmPath,
    WithDecryption: true,
  }

  let params
  try {
    params = await ssm.getParametersByPath(ssmParams).promise()
  } catch (e) {
    throw new Error(e)
  }

  const ssmVals = {}
  params.Parameters.forEach((p) => {
    ssmVals[p.Name.split('/')[3]] = p.Value
  })

  this.config = Object.assign({}, this.defaults, ssmVals) // eslint-disable-line
  // this.config = { ...this.defaults, ...ssmVals } // this, oddly, doesn't compile

  return true
}

module.exports = config
