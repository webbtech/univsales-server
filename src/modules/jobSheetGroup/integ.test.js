import { createTestClient } from 'apollo-server-testing'
import { ApolloServer, gql } from 'apollo-server'

import schema from '../../schema'
import config from '../../config/config'
import db from '../../mongo/connect'

// to test this file use: yarn test:w src/modules/jobSheetGroup/integ.test.js

const { typeDefs, resolvers } = schema
let query

const groupID = '5b195db0c2e75ffe21fdd0ed'

const JOBSHEET_GROUP = gql`
query jobSheetGroup($groupID: ID!) {
  jobSheetGroup(groupID: $groupID) {
    _id
    jobsheetID
    costs {
      __typename
      discounted
      extendTotal
      extendUnit
      install
      installType
      netUnit
      options
      trim
      windows
    }
    dims {
      __typename
      height {
        __typename
        decimal
        diff
        fraction
        inch
      }
      width {
        __typename
        decimal
        diff
        fraction
        inch
      }
    }
    items {
      __typename
      _id
      costs {
        __typename
        extendUnit
        extendTotal
      }
      dims {
        __typename
        height {
          __typename
          decimal
          fraction
          inch
          overSize
          round
          underSize
        }
        width {
          __typename
          decimal
          fraction
          inch
          overSize
          round
          underSize
        }
      }
      product {
        name
      }
      productID
      qty
      specs {
        extendSqft
        overSize
        sqft
      }
    }
    qty
    rooms
    specs {
      groupTypeDescription
      installType
      options
      sqft
      style
      trim
    }
    createdAt
    updatedAt
  }
}`

beforeAll(async () => {
  const cfg = await config.load()
  await db.connect(cfg)

  const server = new ApolloServer({
    typeDefs,
    resolvers,
  })
  const testClient = createTestClient(server)
  query = testClient.query // eslint-disable-line prefer-destructuring
})

test('jobSheetGroup', async () => {
  const res = await query({ query: JOBSHEET_GROUP, variables: { groupID } })
  const { jobSheetGroup } = res.data
  expect(jobSheetGroup).toBeTruthy()
  // expect(res).toMatchSnapshot()
})
