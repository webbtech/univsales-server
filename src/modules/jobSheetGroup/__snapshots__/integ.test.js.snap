// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jobSheetGroup 1`] = `
Object {
  "data": Object {
    "jobSheetGroup": Object {
      "_id": "5b195db0c2e75ffe21fdd0ed",
      "costs": Object {
        "__typename": "JobSheetGroupCosts",
        "discounted": 0,
        "extendTotal": 3450,
        "extendUnit": 3450,
        "install": 1000,
        "installType": 500,
        "netUnit": 3450,
        "options": 150,
        "trim": 100,
        "windows": 1700,
      },
      "createdAt": "1528389040817",
      "dims": Object {
        "__typename": "JobSheetGroupDims",
        "height": Object {
          "__typename": "GroupWindowDims",
          "decimal": 22,
          "diff": null,
          "fraction": "",
          "inch": 22,
        },
        "width": Object {
          "__typename": "GroupWindowDims",
          "decimal": 22,
          "diff": null,
          "fraction": "",
          "inch": 22,
        },
      },
      "items": Array [
        Object {
          "__typename": "JobSheetGroupItem",
          "_id": "5b1a70452aac048edc7ec028",
          "costs": Object {
            "__typename": "GroupItemCosts",
            "extendTotal": 860,
            "extendUnit": 430,
          },
          "dims": Object {
            "__typename": "GroupItemDims",
            "height": Object {
              "__typename": "WindowDims",
              "decimal": 62,
              "fraction": null,
              "inch": 62,
              "overSize": null,
              "round": 62,
              "underSize": null,
            },
            "width": Object {
              "__typename": "WindowDims",
              "decimal": 21,
              "fraction": null,
              "inch": 21,
              "overSize": null,
              "round": 22,
              "underSize": null,
            },
          },
          "product": Object {
            "name": "2 Casements (L & R)",
          },
          "productID": "57855061982d822a04b760a8",
          "qty": 2,
          "specs": Object {
            "extendSqft": 20,
            "overSize": 0,
            "sqft": 10,
          },
        },
        Object {
          "__typename": "JobSheetGroupItem",
          "_id": "5b1a70452aac048f427ec027",
          "costs": Object {
            "__typename": "GroupItemCosts",
            "extendTotal": 840,
            "extendUnit": 280,
          },
          "dims": Object {
            "__typename": "GroupItemDims",
            "height": Object {
              "__typename": "WindowDims",
              "decimal": 62,
              "fraction": null,
              "inch": 62,
              "overSize": null,
              "round": 62,
              "underSize": null,
            },
            "width": Object {
              "__typename": "WindowDims",
              "decimal": 20,
              "fraction": null,
              "inch": 20,
              "overSize": null,
              "round": 20,
              "underSize": null,
            },
          },
          "product": Object {
            "name": "Fixed",
          },
          "productID": "57855061982d822a04b760a4",
          "qty": 3,
          "specs": Object {
            "extendSqft": 27,
            "overSize": 0,
            "sqft": 9,
          },
        },
      ],
      "jobsheetID": "5b1846d52aac0450227ebfe9",
      "qty": 1,
      "rooms": Array [
        "DR",
      ],
      "specs": Object {
        "groupTypeDescription": "Bow Window 2 Casements, 3 Fixed",
        "installType": null,
        "options": "Pebble: Low E & Argon Gas (dark)
Colonial Bars
",
        "sqft": 4,
        "style": null,
        "trim": "B.M. & Interior \\"J\\"
2 3/4\\" F.J. Trim
",
      },
      "updatedAt": "1582647497225",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
